### PR TITLE
[CI] update codecov action to v4

### PR DESCRIPTION
**Pull request by kasia-kujawa**
**Description:** update codecov action to v4

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32259

### CODE REVIEW

**Comment by crobert-1** [comment]
It looks like to add support for v4 we'd need to add a token for this action. ([v4 breaking change reference.](https://github.com/codecov/codecov-action?tab=readme-ov-file#breaking-changes)) 

Workflow changes unfortunately aren't picked up in their respective PR, so this failure wouldn't be hit until this gets merged into main, from my understanding.

**Comment by crobert-1** [comment]
Well it looks like the action actually passed here. I admit I'm a bit confused as to what's going on. I see from the output:
```
info - 2024-04-09 15:49:49,207 -- The PR is happening in a forked repo. Using tokenless upload.
```
So maybe it did pick up your changes, but it's not requiring a token since the PR is a fork? But if this gets merged `main` will fail with the token error, perhaps?

File: .github/workflows/build-and-test.yml [hunk 1]
```diff
@@ -333,7 +333,7 @@ jobs:
       - name: Upload coverage report
         uses: Wandalen/wretry.action@v3.0.1
         with:
-          action: codecov/codecov-action@v3
+          action: codecov/codecov-action@v4
           with: |
             fail_ci_if_error: true
             verbose: true
```
Comment (crobert-1) [inline .github/workflows/build-and-test.yml:339]:
```suggestion
            verbose: true
            token: ${{ secrets.CODECOV_TOKEN }}
```
I think this should enable us to upgrade 👍

**Comment by kkujawa-sumo** [comment]
>Well it looks like the action actually passed here. I admit I'm a bit confused as to what's going on. I see from the output:
info - 2024-04-09 15:49:49,207 -- The PR is happening in a forked repo. Using tokenless upload.
So maybe it did pick up your changes, but it's not requiring a token since the PR is a fork? But if this gets merged main will fail with the token error, perhaps?

@crobert-1  I wasn't sure if token is required for this repo seeing this information
```
PRs made from forks to the upstream public repos will support tokenless (e.g. contributors to OS projects do not need the upstream repo's Codecov token)
```
but definitely it is better to add this. 

I need a help with adding the secret with token for this repo, link to the instruction for generating token is [here](https://github.com/codecov/codecov-action?tab=readme-ov-file#usage).

**Comment by kkujawa-sumo** [comment]
As this is done https://github.com/open-telemetry/community/issues/1926 so I expect that the `CODECOV_TOKEN`  secret exists

**Review by andrzej-stencel — APPROVED** [review]

**Pull request merged**